File: 7.Sorting/HeapSort/HeapSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int left(int i) {
    return 2 * i;
}

int right(int i) {
    return 2 * i + 1;
}

void MAX_HEAP(int arr[], int i, int n) {
    int l = left(i);
    int r = right(i);
    int largest = i;

    if (l <= n && arr[l] > arr[largest])
        largest = l;
    if (r <= n && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        swap(&arr[i], &arr[largest]);
        MAX_HEAP(arr, largest, n);
    }
}

void BuildHeap(int arr[], int n) {
    for (int i = n / 2; i >= 1; i--)
        MAX_HEAP(arr, i, n);
}

void HeapSort(int arr[], int n) {
    BuildHeap(arr, n);
    for (int i = n; i >= 2; i--) {
        swap(&arr[1], &arr[i]);
        MAX_HEAP(arr, 1, i - 1);
    }
}

void display(int arr[], int size) {
    for (int i = 1; i <= size; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int A[100001], n;
    clock_t start, end;

    printf("Enter n (up to 100000): ");
    scanf("%d", &n);

    // Fill array from index 1 for 1-based indexing
    for (int i = 1; i <= n; i++) {
        A[i] = rand() % 10000;
    }

    printf("Original array:\n");
    display(A, n);

    start = clock();
    HeapSort(A, n);
    end = clock();

    printf("Sorted array:\n");
    display(A, n);

    double time_taken = (double)(end - start) / CLOCKS_PER_SEC;
    printf("Time taken to sort: %lf seconds\n", time_taken);

    return 0;
}

File: 7.Sorting/HeapSort/HeapSort.py
import time
import random

def max_heapify(arr, n, i):
    largest = i
    left = 2 * i   
    right = 2 * i + 1

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        max_heapify(arr, n, largest)

def build_heap(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        max_heapify(arr, n, i)

def heap_sort(arr):
    n = len(arr)
    build_heap(arr)
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        max_heapify(arr, i, 0)

def display(arr):
    print(" ".join(map(str, arr)))

n = int(input("Enter n: "))
arr = [random.randint(1, 10000) for _ in range(n)]

print("Original array:")
display(arr)

start = time.time()
heap_sort(arr)
end = time.time()

print("Sorted array:")
display(arr)

print(f"Time taken to sort: {end - start:.6f} seconds")
